### PR TITLE
Add RecordEventsSpanImpl to SDK.

### PR DESCRIPTION
**Pull request by songy23**
Updates #279.

~~This PR is blocked waiting on https://github.com/open-telemetry/opentelemetry-java/pull/354.~~

### CODE REVIEW

File: sdk/src/main/java/io/opentelemetry/sdk/trace/internal/ConcurrentIntrusiveList.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+ *
+ *  {@literal @}Override
+ *   void setPrev(MyClass element) {
+ *     prev = element;
+ *   }
+ * }
+ * </code></pre>
+ */
+@ThreadSafe
+public final class ConcurrentIntrusiveList<T extends Element<T>> {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/internal/ConcurrentIntrusiveList.java]: We don't need this right now. We will implement the active spans as a SpanProcessor, we can discuss that later.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java [hunk 2]
```diff
@@ -0,0 +1,743 @@
...
+  private Status status;
+  // The end time of the span.
+  @GuardedBy("this")
+  private long endNanoTime;
+  // True if the span is ended.
+  @GuardedBy("this")
+  private boolean hasBeenEnded;
+
+  @GuardedBy("this")
+  private boolean sampleToLocalSpanStore;
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: Not needed.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java [hunk 3]
```diff
@@ -0,0 +1,743 @@
...
+  // Contains the identifiers associated with this Span.
+  private final SpanContext context;
+  // The parent SpanId of this span. Null if this is a root span.
+  @Nullable private final SpanId parentSpanId;
+  // True if the parent is on a different process.
+  @Nullable private final Boolean hasRemoteParent;
+  // Active trace configs when the Span was created.
+  private final TraceConfig traceConfig;
+  // Handler called when the span starts and ends.
+  private final StartEndHandler startEndHandler;
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: This is now the SpanProcessor.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java [hunk 4]
```diff
@@ -0,0 +1,743 @@
...
+@ThreadSafe
+public final class RecordEventsSpanImpl implements SpanSdk, Element<RecordEventsSpanImpl> {
+  private static final Logger logger = Logger.getLogger(Tracer.class.getName());
+
+  // Contains the identifiers associated with this Span.
+  private final SpanContext context;
+  // The parent SpanId of this span. Null if this is a root span.
+  @Nullable private final SpanId parentSpanId;
+  // True if the parent is on a different process.
+  @Nullable private final Boolean hasRemoteParent;
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: This has been removed.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java [hunk 5]
```diff
@@ -0,0 +1,743 @@
...
+  // Active trace configs when the Span was created.
+  private final TraceConfig traceConfig;
+  // Handler called when the span starts and ends.
+  private final StartEndHandler startEndHandler;
+  // The displayed name of the span.
+  @GuardedBy("this")
+  private String name;
+  // The kind of the span.
+  @Nullable private final Kind kind;
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: No longer nullable, default is INTERNAL.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java [hunk 6]
```diff
@@ -0,0 +1,743 @@
...
+  @Nullable
+  private AttributesWithCapacity attributes;
+  // List of recorded events.
+  @GuardedBy("this")
+  @Nullable
+  private TraceEvents<TimedEvent> events;
+  // List of recorded links to parent and child spans.
+  @GuardedBy("this")
+  @Nullable
+  private TraceEvents<Link> links;
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: Is this TraceEvents<Links>???

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java [hunk 7]
```diff
@@ -0,0 +1,743 @@
...
+
+    // It is called after each put or putAll call in order to determine if the eldest inserted
+    // entry should be removed or not.
+    @Override
+    protected boolean removeEldestEntry(Map.Entry<String, AttributeValue> eldest) {
+      return size() > this.capacity;
+    }
+  }
+
+  private static final class TraceEvents<T> {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: Consider to avoid an allocation and embed this directly into the Span:

```java
private int totalRecordedEvents;
private EvictingQueue<Events> events;
private int totalRecordedLinks;
private EvictingQueue<Link> links;
```

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java (hunk 4 above)
Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: It's still being used in the proto: https://github.com/open-telemetry/opentelemetry-java/blob/ca0ec9dfa83fd8c4b1a9e4a5af3db75cf64c6cbc/sdk/src/main/proto/trace/v1/trace.proto#L240-L245.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java (hunk 3 above)
Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: I see, updated.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java (hunk 5 above)
Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: Done.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java (hunk 6 above)
Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: `TraceEvents` is just an internal helper class used to track the dropped count. This is copied from https://github.com/census-instrumentation/opencensus-java/blob/282d742f64c4ffd2273608d644bffbefe538afdd/impl_core/src/main/java/io/opencensus/implcore/trace/RecordEventsSpanImpl.java#L92-L95.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java (hunk 2 above)
Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: Removed.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java (hunk 7 above)
Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: SG, updated.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/internal/ConcurrentIntrusiveList.java (hunk 1 above)
Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/internal/ConcurrentIntrusiveList.java]: Got it, removed for now.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+import io.opentelemetry.trace.SpanContext;
+import io.opentelemetry.trace.SpanId;
+import io.opentelemetry.trace.Status;
+import io.opentelemetry.trace.TraceId;
+import io.opentelemetry.trace.Tracestate;
+import java.nio.charset.Charset;
+import java.util.Map;
+
+// Utilities to convert to trace protos.
+final class TraceProtoUtils {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java]: Please add unit tests for this class.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+  static io.opentelemetry.proto.resource.v1.Resource toProtoResource(Resource resource) {
+    io.opentelemetry.proto.resource.v1.Resource.Builder builder =
+        io.opentelemetry.proto.resource.v1.Resource.newBuilder();
+    for (Map.Entry<String, String> entry : resource.getLabels().entrySet()) {
+      builder.putLabels(entry.getKey(), entry.getValue());
+    }
+    return builder.build();
+  }
+
+  static SpanKind toProtoKind(Kind kind) {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java]: See https://github.com/open-telemetry/opentelemetry-proto/issues/8

File: sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    builder.setTime(converter.convertNanoTime(timedEvent.getNanotime()));
+    builder.setEvent(
+        io.opentelemetry.proto.trace.v1.Span.TimedEvent.Event.newBuilder()
+            .setName(TruncatableString.newBuilder().setValue(timedEvent.getName()).build())
+            .setAttributes(toProtoAttributes(timedEvent.getAttributes()))
+            .build());
+    return builder.build();
+  }
+
+  static Links toProtoLinks(EvictingQueue<Link> links, int totalRecordedLinks) {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java]: Collection<Link> is more generic.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+        break;
+      case STRING:
+        builder.setStringValue(
+            TruncatableString.newBuilder().setValue(attributeValue.getStringValue()).build());
+    }
+    return builder.build();
+  }
+
+  static TimedEvents toProtoTimedEvents(
+      EvictingQueue<TimedEvent> events, int totalRecordedEvents, TimestampConverter converter) {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java]: Collection<TimedEvent> is more generic.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    switch (kind) {
+      case CLIENT:
+        return SpanKind.CLIENT;
+      case SERVER:
+        return SpanKind.SERVER;
+    }
+    return SpanKind.UNRECOGNIZED;
+  }
+
+  static Attributes toProtoAttributes(AttributesWithCapacity attributes) {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java]: static Attributes toProtoAttributes(Map<String, AttributeValue> attributes, int droppedAttributes); This way it can be unified with the next method.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java (hunk 8 above)
Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java]: I'll post a separate PR on proto utils and tests.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java (hunk 9 above)
Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java]: SG, I'll update the proto first.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java (hunk 12 above)
Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/TraceProtoUtils.java]: SG, will do in the next PR.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java [hunk 13]
```diff
@@ -0,0 +1,510 @@
...
+    }
+  }
+
+  /**
+   * Returns the end nano time (see {@link System#nanoTime()}). If the current {@code Span} is not
+   * ended then returns {@link Clock#nowNanos()}.
+   *
+   * @return the end nano time.
+   */
+  public long getEndNanoTime() {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: Why public? Same question for every method that is not an override?

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java [hunk 14]
```diff
@@ -0,0 +1,510 @@
...
+  public io.opentelemetry.proto.trace.v1.Span toSpanProto() {
+    throw new UnsupportedOperationException("to be implemented");
+  }
+
+  /**
+   * Returns the status of the {@code Span}. If not set defaults to {@link Status#OK}.
+   *
+   * @return the status of the {@code Span}.
+   */
+  public Status getStatus() {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: Why public?

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java [hunk 15]
```diff
@@ -0,0 +1,510 @@
...
+    }
+  }
+
+  /**
+   * Returns the latency of the {@code Span} in nanos. If still active then returns now() - start
+   * time.
+   *
+   * @return the latency of the {@code Span} in nanos.
+   */
+  public long getLatencyNs() {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: ditto

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java [hunk 16]
```diff
@@ -0,0 +1,510 @@
...
+      return hasBeenEnded ? endNanoTime - startNanoTime : clock.nowNanos() - startNanoTime;
+    }
+  }
+
+  /**
+   * Returns the kind of this {@code Span}.
+   *
+   * @return the kind of this {@code Span}.
+   */
+  public Kind getKind() {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: ditto.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java [hunk 17]
```diff
@@ -0,0 +1,510 @@
...
+
+  /**
+   * Returns an immutable representation of all the data from this {@code Span}.
+   *
+   * <p>This should only be called with an ended {@code Span}.
+   *
+   * @return an immutable representation of all the data from this {@code Span}.
+   * @throws IllegalStateException if the Span doesn't have RECORD_EVENTS option.
+   */
+  public SpanData toSpanData() {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: Why is this here? I thought that we use the proto class to expose the format, also why public?

**Comment by songy23** [comment]
PTAL.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java [hunk 18]
```diff
@@ -0,0 +1,510 @@
...
+      if (hasBeenEnded) {
+        logger.log(Level.FINE, "Calling addEvent() on an ended Span.");
+        return;
+      }
+      getInitializedEvents().add(timedEvent);
+      totalRecordedEvents++;
+    }
+  }
+
+  private static List<SpanData.TimedEvent> toSpanDataTimedEvents(
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: why do we have all the to SpanData methods?

**Comment by bogdandrutu** [comment]
If I was not clear we still want to use the toProto and SpanProto class to expose the data but only all the conversion methods that you had to be in a separate PR with tests.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java (hunk 14 above)
Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: Changed to package-private.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java (hunk 13 above)
Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: Updated.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java (hunk 17 above)
Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: Removed this method for now.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java (hunk 18 above)
Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: Removed.

**Comment by bogdandrutu** [comment]
Keeping the SpanData in the API is independent of this change. For the events we record we want to expose them as the Span proto. If we keep the SpanData we will have a way to transform that into the proto as well since that is a different source of data but from consumer perspective we always give them the proto

File: sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java [hunk 19]
```diff
@@ -0,0 +1,456 @@
...
+  @GuardedBy("this")
+  private EvictingQueue<TimedEvent> getInitializedEvents() {
+    if (events == null) {
+      events = EvictingQueue.<TimedEvent>create((int) traceConfig.getMaxNumberOfEvents());
+    }
+    return events;
+  }
+
+  @GuardedBy("this")
+  private EvictingQueue<SpanData.Link> getInitializedLinks() {
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: Why do we not use Link interface directly here?

Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: Because `Link` is an interface and we need a concrete implementation of it to implement `addLink(SpanContext spanContext)` and `addLink(SpanContext spanContext, Map<String, AttributeValue> attributes)`. If we don't use `SpanData.Link` here, we need to add a `LinkImpl` in the SDK. WDYT?

Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: But the list of Links can still be the interface. When we add to the list we add the SpanData.Link

Comment (songy23) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java]: I got your point - updated in https://github.com/open-telemetry/opentelemetry-java/commit/d3ac6dc171f04a8e5cea0b4995134c7fc8e8fd36.
